src/stirling/source_connectors/socket_tracer/protocols/cql/stitcher.cc

```diff
@@ -372,77 +372,120 @@ StatusOr<Record> ProcessSolitaryResp(Frame* resp_frame) {
 
-  for (auto& resp_frame : *resp_frames) {
-    bool found_match = false;
-
-    // Event responses are special: they have no request.
-    if (resp_frame.hdr.opcode == Opcode::kEvent) {
-      StatusOr<Record> record_status = ProcessSolitaryResp(&resp_frame);
-      if (record_status.ok()) {
-        entries.push_back(record_status.ConsumeValueOrDie());
-      } else {
-        VLOG(1) << record_status.msg();
-        ++error_count;
-      }
-      continue;
-    }
-
-    // Search for matching req frame
-    for (auto& req_frame : *req_frames) {
-      if (resp_frame.hdr.stream == req_frame.hdr.stream) {
-        VLOG(2) << absl::Substitute("req_op=$0 msg=$1", magic_enum::enum_name(req_frame.hdr.opcode),
-                                    req_frame.msg);
-
-        StatusOr<Record> record_status = ProcessReqRespPair(&req_frame, &resp_frame);
+  // iterate through all deques of responses associated with a specific streamID and find the
+  // matching request
+  for (auto it = responses->begin(); it != responses->end(); it++) {
+    cass::stream_id stream_id = it->first;
+    std::deque<cass::Frame>& resp_frames = it->second;
+
+    bool kEventHandled = false;
+    for (cass::Frame& resp_frame : resp_frames) {
+      // Event responses are special: they have no request.
+      if (resp_frame.hdr.opcode == Opcode::kEvent) {
+        kEventHandled = true;
+        StatusOr<Record> record_status = ProcessSolitaryResp(&resp_frame);
         if (record_status.ok()) {
           entries.push_back(record_status.ConsumeValueOrDie());
         } else {
           VLOG(1) << record_status.ToString();
           ++error_count;
         }
+      }
+    }
 
-        // Found a match, so remove both request and response.
-        // We don't remove request frames on the fly, however,
-        // because it could otherwise cause unnecessary churn/copying in the deque.
-        // This is due to the fact that responses can come out-of-order.
-        // Just mark the request as consumed, and clean-up when they reach the head of the queue.
-        // Note that responses are always head-processed, so they don't require this optimization.
-        found_match = true;
-        req_frame.consumed = true;
-        break;
+    auto pos = requests->find(stream_id);
+    if (pos == requests->end()) {
+      VLOG(1) << absl::Substitute("Could not find any requests for stream = $0", stream_id);
+      // if we don't find a matching request, we can't do anything with this response
+      // so clean it up
+      resp_frames.clear();
+      if (!kEventHandled) {
+        ++error_count;
       }
+      continue;
+    }
+
+    // we found a potential set of requests for this stream ID
+    std::deque<cass::Frame>& req_frames = pos->second;
+    std::deque<uint64_t> req_timestamps = std::deque<uint64_t>();
+    // get just the timestamps for matching with responses
+    for (cass::Frame& req_frame : req_frames) {
+      req_timestamps.push_back(req_frame.timestamp_ns);
```

Comment:
There should be a way to run the timestamp filtering algorithm without building a separate timestamps dequeue. Also (have not thought about this...) does the timestamp dequeue need to be a dequeue. Can we pre-allocate a vector here?

Reply:
Good point, I've updated the logic to directly extract the timestamp from req_frames during search via a lambda comparator function passed to `upper_bound`.